cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,206 @@
+	char *executable;
+
+	// Flag indicating that --version was passed on command line.
+	bool is_version_query;
+	// Flag indicating that --classic was passed on command line.
+	bool is_classic_confinement;
+};
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc or argv are NULL");
+		goto out;
+	}
+	// Use local copies of argc and argv.
+	int argc = *argcp;
+	char **argv = *argvp;
+
+	if (argc == 0) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc is zero");
+		goto out;
+	}
+
+	args = calloc(1, sizeof *args);
+	if (args == NULL) {
+		die("cannot allocate memory for command line arguments object");
+	}
+	// Check if we're being called through the ubuntu-core-launcher symlink.
+	// When this happens we want to skip the first positional argument as it is
+	// the security tag repeated (legacy).
+	bool ignore_first_tag = false;
+	char *basename = strrchr(argv[0], '/');
+	if (basename != NULL) {
+		// NOTE: this is safe because we, at most, may move to the NUL byte
+		// that compares to an empty string.
+		basename += 1;
+		if (strcmp(basename, "ubuntu-core-launcher") == 0) {
+			ignore_first_tag = true;
+		}
+	}
+	// Parse arguments that we've got
+	int optind;
+	for (optind = 1; optind < argc; ++optind) {
+		// Look at all the options switches that start with the minus sign ('-')
+		if (argv[optind][0] == '-') {
+			// Handle option switches
+			if (strcmp(argv[optind], "--version") == 0) {
+				args->is_version_query = true;
+				// NOTE: --version short-circuits the parser to finish
+				break;
+			} else if (strcmp(argv[optind], "--classic") == 0) {
+				args->is_classic_confinement = true;
+			} else {
+				// Report unhandled option switches
+				err =
+				    sc_error_init(SC_ARGS_DOMAIN,
+						  SC_ARGS_ERR_USAGE,
+						  "unrecognized command line option: %s",
+						  argv[optind]);
+				goto out;
+			}
+		} else {
+			// Handle positional arguments
+			if (args->security_tag == NULL) {
+				// The first positional argument (that is not an option switch)
+				// becomes the security tag.
+				if (ignore_first_tag) {
+					// Unless we are called as ubuntu-core-launcher, then we
+					// just swallow and ignore that security tag altogether.
+					ignore_first_tag = false;
+					continue;
+				}
+				args->security_tag = strdup(argv[optind]);
+				if (args->security_tag == NULL) {
+					die("cannot allocate memory for security tag");
+				}
+			} else if (args->executable == NULL) {
+				// The second positional argument becomes the executable name.
+				args->executable = strdup(argv[optind]);
+				if (args->executable == NULL) {
+					die("cannot allocate memory for executable name");
+				}
+				// No more positional arguments are required.
```

Comment:
Good point, we don't need mixed flags/positionals in any way. I'll simplify the code to handle this.